src/serializer/ResidualHeapVisitor.js

```diff
@@ -487,30 +495,90 @@ export class ResidualHeapVisitor {
         this.referencedDeclaredValues.add(value);
       },
       recordDelayedEntry: (entry: GeneratorEntry) => {
-        this.delayedVisitGeneratorEntries.push({ generator, entry });
+        this.delayedVisitGeneratorEntries.push({ baseContext, generator, entry });
       },
     };
   }
 
-  visitGenerator(generator: Generator): void {
+  visitGenerator(generator: Generator, afterGeneratorVisit?: () => void): void {
     this._withScope(generator, () => {
-      generator.visit(this.createGeneratorVisitCallbacks(generator));
+      generator.visit(this.createGeneratorVisitCallbacks(generator, this.baseContext));
+      if (afterGeneratorVisit) afterGeneratorVisit();
     });
   }
 
-  visitRoots(): void {
-    this.visitGenerator(this.realmGenerator);
+  visitRoots(additionalFunctionValuesAndEffects: Map<FunctionValue, Effects>): void {
+    let generator = this.realm.generator;
+    invariant(generator);
+    this.visitGenerator(generator);
     for (let [, moduleValue] of this.modules.initializedModules) this.visitValue(moduleValue);
-    // Do a fixpoint over all pure generator entries to make sure that we visit
-    // arguments of only BodyEntries that are required by some other residual value
-    let oldDelayedEntries = [];
-    while (oldDelayedEntries.length !== this.delayedVisitGeneratorEntries.length) {
-      oldDelayedEntries = this.delayedVisitGeneratorEntries;
-      this.delayedVisitGeneratorEntries = [];
-      for (let { generator, entry } of oldDelayedEntries)
-        this._withScope(generator, () => {
-          generator.visitEntry(entry, this.createGeneratorVisitCallbacks(generator));
-        });
-    }
+    this.realm.evaluateAndRevertInGlobalEnv(() => {
+      if (additionalFunctionValuesAndEffects.size) {
+        for (let [functionValue, effects] of additionalFunctionValuesAndEffects.entries()) {
+          let [r, g, ob, pb: Map<PropertyBinding, void | Descriptor>, co] = effects;
+          // Need to do this fixup because otherwise we will skip over this function's
+          // generator in the _getTarget scope lookup
+          g.parent = functionValue.parent;
+          functionValue.parent = g;
+          // result -- ignore TODO: return the result from the function somehow
+          // Generator -- visit all entries
+          // Bindings -- only need to serialize bindings if they're captured by some nested function ??
+          //          -- need to apply them and maybe need to revisit functions in ancestors to make sure
+          //          -- we don't overwrite anything they capture
+          //          -- TODO: deal with these properly
+          // PropertyBindings -- visit any property bindings that aren't to createdobjects
+          // CreatedObjects -- should take care of itself
+          this.realm.applyEffects([r, new Generator(this.realm), ob, pb, co]);
```

Comment:
High-level, ResidualFuncitonVisitor should be a class doing pure tagging for values in generator/heap tree. It's weird to call applyEffects() here, maybe move this into functions.js in evaluation stage.

Reply:
I can't avoid calling applyEffects -- the point is to be able to be able to tag things in Effects that will be serialized for additional functions (they will get applied again in serialization). I also can't apply everything upfront because I need to know what changes belong to what additional function.